File: packetCapture/src/main/java/com/b_knabe/packet_capture/core/tunnels/UdpTunnel.java
package com.b_knabe.packet_capture.core.tunnels;


import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.b_knabe.packet_capture.core.proxy_servers.UdpProxyServer;
import com.b_knabe.packet_capture.core.nat.NatSession;
import com.b_knabe.packet_capture.core.nat.NatSessionManager;
import com.b_knabe.packet_capture.core.util.common.ACache;
import com.b_knabe.packet_capture.core.util.common.IOUtils;
import com.b_knabe.packet_capture.core.util.common.ThreadPool;
import com.b_knabe.packet_capture.core.util.common.TimeFormatter;
import com.b_knabe.packet_capture.core.util.net_utils.tcp.TcpDataSaver;
import com.b_knabe.packet_capture.core.util.process_parse.PortSessionInfoService;
import com.b_knabe.packet_capture.core.vpn.VpnProxyServer;
import com.b_knabe.packet_capture.tcp_ip_level.Packet;

import java.io.File;
import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

public class UdpTunnel implements KeyHandler {
    private static final String TAG = "UdpTunnel";
    private static final int HEADER_SIZE = Packet.IP4_HEADER_SIZE + Packet.UDP_HEADER_SIZE;
    private final Selector selector;
    private final UdpProxyServer udpProxyServer;
    private final Queue<Packet> outputQueue;
    private final ConcurrentLinkedQueue<Packet> toNetWorkPackets = new ConcurrentLinkedQueue<>();
    private final NatSession session;
    private final Handler handler;
    private String ipAndPort;
    private TcpDataSaver tcpDataSaver;
    private Packet referencePacket;
    private SelectionKey selectionKey;
    private DatagramChannel channel;
    private Short portKey;

    public UdpTunnel(Selector selector, UdpProxyServer udpProxyServer, Packet packet, Queue<Packet> outputQueue, short portKey) {
        this.selector = selector;
        this.udpProxyServer = udpProxyServer;
        this.referencePacket = packet;
        ipAndPort = packet.getIpAndPort();
        this.outputQueue = outputQueue;
        this.portKey = portKey;
        session = NatSessionManager.getSession(portKey);
        handler = new Handler(Looper.getMainLooper());
        String dir = new StringBuilder()
                .append(TcpDataSaver.DATA_DIR)
                .append(TimeFormatter.formatToYYMMDDHHMMSS(session.vpnStartTime))
                .append("/")
                .append(session.getUniqueName())
                .toString();
        tcpDataSaver = new TcpDataSaver(dir);
    }


    private void processKey(SelectionKey key) {
        if (key.isWritable()) {
            processSend();
        } else if (key.isReadable()) {
            processReceived();
        }
        updateInterests();
    }

    private void processReceived() {
        ByteBuffer receiveBuffer = ByteBuffer.allocate(VpnProxyServer.getMtu());
        // Leave space for the header
        receiveBuffer.position(HEADER_SIZE);
        int readBytes;
        try {
            readBytes = channel.read(receiveBuffer);
        } catch (Exception e) {
            udpProxyServer.closeUdpTunnel(this);
            return;
        }
        if (readBytes == -1) {
            udpProxyServer.closeUdpTunnel(this);
        } else if (readBytes == 0) {
            Log.d(TAG, "read no data :" + ipAndPort);
        } else {
            Log.d(TAG, "read readBytes:" + readBytes + " ipAndPort:" + ipAndPort);
            Packet newPacket = referencePacket.duplicated();
            newPacket.updateUDPBuffer(receiveBuffer, readBytes);
            receiveBuffer.position(HEADER_SIZE + readBytes);
            outputQueue.offer(newPacket);
            session.receivePacketNum++;
            session.receiveByteNum += readBytes;
            session.lastRefreshTime = System.currentTimeMillis();
            if (tcpDataSaver != null) {
                saveData(receiveBuffer.array(), readBytes, false);
            }
        }
    }

    private void saveData(byte[] array, int saveSize, boolean isRequest) {
        TcpDataSaver.TcpData saveTcpData = new TcpDataSaver.TcpData
                .Builder()
                .offSet(HEADER_SIZE)
                .length(saveSize)
                .needParseData(array)
                .isRequest(isRequest)
                .build();
        tcpDataSaver.addData(saveTcpData);
    }

    private void processSend() {
        Log.d(TAG, "processWriteUDPData " + ipAndPort);
        Packet toNetWorkPacket = getToNetWorkPackets();
        if (toNetWorkPacket == null) {
            Log.d(TAG, "write data  no packet ");
            return;
        }
        try {
            ByteBuffer payloadBuffer = toNetWorkPacket.backingBuffer;
            session.packetSent++;
            int sendSize = payloadBuffer.limit() - payloadBuffer.position();
            session.bytesSent += sendSize;
            if (tcpDataSaver != null) {
                saveData(payloadBuffer.array(), sendSize, true);
            }
            session.lastRefreshTime = System.currentTimeMillis();
            while (payloadBuffer.hasRemaining()) {
                channel.write(payloadBuffer);
            }
        } catch (IOException e) {
            Log.w(TAG, "Network write error: " + ipAndPort, e);
            udpProxyServer.closeUdpTunnel(this);
        }
    }

    public void initConnection() {
        Log.d(TAG, "init  ipAndPort:" + ipAndPort);
        InetAddress destinationAddress = referencePacket.ip4Header.destinationAddress;
        int destinationPort = referencePacket.udpHeader.destinationPort;
        try {
            channel = DatagramChannel.open();
            VpnProxyServer.protect(channel.socket());
            channel.configureBlocking(false);
            channel.connect(new InetSocketAddress(destinationAddress, destinationPort));
            selector.wakeup();
            selectionKey = channel.register(selector,
                    SelectionKey.OP_READ, this);
        } catch (IOException e) {
            IOUtils.close(channel);
            return;
        }
        referencePacket.swapSourceAndDestination();
        addToNetWorkPacket(referencePacket);
    }

    public void processPacket(Packet packet) {
        addToNetWorkPacket(packet);
        updateInterests();
    }

    public void close() {
        try {
            if (selectionKey != null) {
                selectionKey.cancel();
            }
            if (channel != null) {
                channel.close();
            }
            if (session.applicationInfo == null && PortSessionInfoService.getInstance() != null) {
                PortSessionInfoService.getInstance().refreshSessionInfo();
            }
            // Need to delay one second before saving and wait until the app information is completely refreshed
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    ThreadPool.getInstance().execute(new Runnable() {
                        @Override
                        public void run() {
                            if (session.receiveByteNum == 0 && session.bytesSent == 0) {
                                return;
                            }
                            String configFileDir = TcpDataSaver.CONFIG_DIR
                                    + TimeFormatter.formatToYYMMDDHHMMSS(session.vpnStartTime);
                            File parentFile = new File(configFileDir);
                            if (!parentFile.exists()) {
                                parentFile.mkdirs();
                            }
                            //Said it has been saved
                            File file = new File(parentFile, session.getUniqueName());
                            if (file.exists()) {
                                return;
                            }
                            ACache configACache = ACache.get(parentFile);
                            configACache.put(session.getUniqueName(), session);
                        }
                    });
                }
            }, 1000);
        } catch (Exception e) {
            Log.w(TAG, "error to close UDP channel IpAndPort" + ipAndPort + ",error is " + e.getMessage());
        }

    }


    public Packet getToNetWorkPackets() {
        return toNetWorkPackets.poll();
    }

    private void addToNetWorkPacket(Packet packet) {
        toNetWorkPackets.offer(packet);
        updateInterests();
    }

    DatagramChannel getChannel() {
        return channel;
    }

    private void updateInterests() {
        int ops;
        if (toNetWorkPackets.isEmpty()) {
            ops = SelectionKey.OP_READ;
        } else {
            ops = SelectionKey.OP_WRITE | SelectionKey.OP_READ;
        }
        selector.wakeup();
        selectionKey.interestOps(ops);
    }

    public Packet getReferencePacket() {
        return referencePacket;
    }

    @Override
    public void onKeyReady(SelectionKey key) {
        processKey(key);
    }

    public Short getPortKey() {
        return portKey;
    }
}

File: packetCapture/src/main/java/com/ninpou/packetcapture/core/tunnel/UdpTunnel.kt
package com.ninpou.packetcapture.core.tunnel

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.ninpou.packetcapture.core.forward.UdpProxyServer
import com.ninpou.packetcapture.core.nat.NatSession
import com.ninpou.packetcapture.core.nat.NatSessionManager
import com.ninpou.packetcapture.core.util.android.PortHostService
import com.ninpou.packetcapture.core.util.common.ACache
import com.ninpou.packetcapture.core.util.common.ThreadPool
import com.ninpou.packetcapture.core.util.common.TimeFormatter.formatToYYMMDDHHMMSS
import com.ninpou.packetcapture.core.util.net.TcpDataSaver
import com.ninpou.packetcapture.core.vpn.VpnServiceProxy.mtu
import com.ninpou.packetcapture.core.vpn.VpnServiceProxy.protect
import com.ninpou.packetcapture.struct.Packet
import top.srsea.torque.common.IOUtils
import java.io.File
import java.io.IOException
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import java.nio.channels.DatagramChannel
import java.nio.channels.SelectionKey
import java.nio.channels.Selector
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue

class UdpTunnel(private val selector: Selector, private val udpProxyServer: UdpProxyServer, val referencePacket: Packet, outputQueue: Queue<Packet>, portKey: Short) : KeyHandler {
    private val outputQueue: Queue<Packet>
    private val toNetWorkPackets = ConcurrentLinkedQueue<Packet>()
    private val session: NatSession
    private val handler: Handler
    private val ipAndPort: String
    private val tcpDataSaver: TcpDataSaver?
    private var selectionKey: SelectionKey? = null
    var channel: DatagramChannel? = null
        private set
    val portKey: Short
    private fun processKey(key: SelectionKey?) {
        if (key!!.isWritable) {
            processSend()
        } else if (key.isReadable) {
            processReceived()
        }
        updateInterests()
    }

    private fun processReceived() {
        val receiveBuffer = ByteBuffer.allocate(mtu)
        receiveBuffer.position(HEADER_SIZE)
        val readBytes: Int
        readBytes = try {
            channel!!.read(receiveBuffer)
        } catch (e: Exception) {
            udpProxyServer.closeUdpTunnel(this)
            return
        }
        if (readBytes == -1) {
            udpProxyServer.closeUdpTunnel(this)
        } else if (readBytes == 0) {
            Log.d(TAG, "read no data :$ipAndPort")
        } else {
            Log.d(TAG, "read readBytes:$readBytes ipAndPort:$ipAndPort")
            val newPacket = referencePacket.duplicated()
            newPacket.updateUDPBuffer(receiveBuffer, readBytes)
            receiveBuffer.position(HEADER_SIZE + readBytes)
            outputQueue.offer(newPacket)
            session.receivePacketNum++
            session.receiveByteNum += readBytes.toLong()
            session.lastRefreshTime = System.currentTimeMillis()
            if (tcpDataSaver != null) {
                saveData(receiveBuffer.array(), readBytes, false)
            }
        }
    }

    private fun saveData(array: ByteArray, saveSize: Int, isRequest: Boolean) {
        val saveTcpData = TcpDataSaver.TcpData.Builder()
                .offSet(HEADER_SIZE)
                .length(saveSize)
                .needParseData(array)
                .isRequest(isRequest)
                .build()
        tcpDataSaver!!.addData(saveTcpData)
    }

    private fun processSend() {
        Log.d(TAG, "processWriteUDPData $ipAndPort")
        val toNetWorkPacket = getToNetWorkPackets()
        if (toNetWorkPacket == null) {
            Log.d(TAG, "write data  no packet ")
            return
        }
        try {
            val payloadBuffer = toNetWorkPacket.backingBuffer
            session.packetSent++
            val sendSize = payloadBuffer.limit() - payloadBuffer.position()
            session.bytesSent += sendSize
            if (tcpDataSaver != null) {
                saveData(payloadBuffer.array(), sendSize, true)
            }
            session.lastRefreshTime = System.currentTimeMillis()
            while (payloadBuffer.hasRemaining()) {
                channel!!.write(payloadBuffer)
            }
        } catch (e: IOException) {
            Log.w(TAG, "Network write error: $ipAndPort", e)
            udpProxyServer.closeUdpTunnel(this)
        }
    }

    fun initConnection() {
        Log.d(TAG, "init  ipAndPort:$ipAndPort")
        val destinationAddress = referencePacket.ip4Header.destinationAddress
        val destinationPort = referencePacket.udpHeader.destinationPort
        try {
            //? NULL
            channel = DatagramChannel.open()
            protect(channel?.socket())
            channel?.configureBlocking(false)
            channel?.connect(InetSocketAddress(destinationAddress, destinationPort))
            selector.wakeup()
            selectionKey = channel!!.register(selector,
                    SelectionKey.OP_READ, this)
        } catch (e: IOException) {
            IOUtils.close(channel)
            return
        }
        referencePacket.swapSourceAndDestination()
        addToNetWorkPacket(referencePacket)
    }

    fun processPacket(packet: Packet) {
        addToNetWorkPacket(packet)
        updateInterests()
    }

    fun close() {
        try {
            if (selectionKey != null) {
                selectionKey!!.cancel()
            }
            if (channel != null) {
                channel!!.close()
            }
            if (session.appInfo == null && PortHostService.instance != null) {
                PortHostService.instance!!.refreshSessionInfo()
            }
            handler.postDelayed(Runnable {
                ThreadPool.instance.execute(Runnable {
                    if (session.receiveByteNum == 0L && session.bytesSent == 0) {
                        return@Runnable
                    }
                    val configFileDir = (TcpDataSaver.CONFIG_DIR
                            + formatToYYMMDDHHMMSS(session.vpnStartTime))
                    val parentFile = File(configFileDir)
                    if (!parentFile.exists()) {
                        parentFile.mkdirs()
                    }
                    val file = File(parentFile, session.uniqueName)
                    if (file.exists()) {
                        return@Runnable
                    }
                    val configACache = ACache.get(parentFile)
                    configACache.put(session.uniqueName, session)
                })
            }, 1000)
        } catch (e: Exception) {
            Log.w(TAG, "error to close UDP channel IpAndPort" + ipAndPort + ",error is " + e.message)
        }
    }

    fun getToNetWorkPackets(): Packet? {
        return toNetWorkPackets.poll()
    }

    private fun addToNetWorkPacket(packet: Packet) {
        toNetWorkPackets.offer(packet)
        updateInterests()
    }

    private fun updateInterests() {
        val ops: Int
        ops = if (toNetWorkPackets.isEmpty()) {
            SelectionKey.OP_READ
        } else {
            SelectionKey.OP_WRITE or SelectionKey.OP_READ
        }
        selector.wakeup()
        selectionKey!!.interestOps(ops)
    }

    override fun onKeyReady(key: SelectionKey?) {
        processKey(key)
    }

    companion object {
        private const val TAG = "UdpTunnel"
        private const val HEADER_SIZE = Packet.IP4_HEADER_SIZE + Packet.UDP_HEADER_SIZE
    }

    init {
        ipAndPort = referencePacket.ipAndPort
        this.outputQueue = outputQueue
        this.portKey = portKey
        session = NatSessionManager.getSession(portKey)
        handler = Handler(Looper.getMainLooper())
        val dir = StringBuilder()
                .append(TcpDataSaver.DATA_DIR)
                .append(formatToYYMMDDHHMMSS(session.vpnStartTime))
                .append("/")
                .append(session.uniqueName)
                .toString()
        tcpDataSaver = TcpDataSaver(dir)
    }
}